src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -1213,11 +1213,12 @@ ActionsDAGPtr SelectQueryExpressionAnalyzer::appendPrewhere(
         /// Reasons:
         /// 1. Remove remove source columns which are used only in prewhere actions during prewhere actions execution.
         ///    Example: select A prewhere B > 0. B can be removed at prewhere step.
-        /// 2. Store side columns which were calculated during prewhere actions execution if they are used.
-        ///    Example: select F(A) prewhere F(A) > 0. F(A) can be saved from prewhere step.
+        /// 2. Do NOT store side columns which were calculated during prewhere actions execution if they are used.
```

Comment:
There can be one very common performance degradation after this patch, that should be fixed by #54535, though perf tests does not shows anything.

Reply:
> Not sure that I got this question. Can you elaborate?

I'm trying to understand when it is actually necessary to drop these columns. for example, if the query doesn't contain `final` - should we? looks like no. if the expression we calculated doesn't contain stateful functions - again, I'm not quite understand how result will change after merging.

Reply:
I'm sorry for repeating the same questions, but I still not quite get it.

> As you can see in the test I have min/max - they are not stateful.

as I understand it is. it maintains the current min/max and updates it with values from new rows. the problem happens if some of these rows that updated the maximum will then be discarded by final merge or whatever else. so they changed the state while they shouldn't. if there is no state to change, like we have `plus(x, 1)` - how problem could happen?

Reply:
Oh, sorry that I did not get you at first, I see, you was talking about regular, non aggregating functions.

Let's just clarify where this function will be met - in the query, since in the columns you have only `SimpleAggregateFunction` and `AggregateFunction`.

AFAICS even with such functions, there are problems, because you cannot be sure which one will be preferred after applying `FINAL`.

Let's consider `CollapsingMergeTree`, for column `x' you have two values (only one of which will be left after applying `FINAL`) for the primary key: `X1` and `X2`, and non aggregating function `func` in query, such as, `func(X1) = X1'` and `func(X2) = X2'`, and you have a query like `select * from table final prewhere func(x) = X1'` but after applying `FINAL` the merging algorithm can leave `X2'`, and so the cache execution of `func` will be incorrect for this row.

Does this make sense?

_P.S. And the newly added test contains non aggregate function in query - `assumeNotNull`_

Reply:
Looking into this deeper and it seems that this is not true for non-aggregating columns. And it is only possible for `SimpleAggregateFunction`. I've adjusted the patch (and description).

Reply:
and not for normal `AggregateFunction`?

Reply:
You cannot write expression against AggregateFunction, since it stores the state (well, you can, but who will do this?)